Add tool registry for judges

dbczumar opened the pull request:
### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

Add tool registry for judges

### How is this PR tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [X] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- patch -->

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [X] No (this PR will be included in the next minor release)

dbczumar committed:
fix

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Add comprehensive test coverage for judge tool registry

- Add test_judge_tool_registry.py with 7 tests covering registry functionality
  - Tool registration, listing, and invocation
  - Error handling for missing tools, invalid JSON, invalid arguments
  - Global API functions (register_judge_tool, list_judge_tools, invoke_judge_tool)
  - Auto-registration of builtin tools
- Add test_judge_tool_get_trace_info.py with 6 tests covering GetTraceInfoTool
  - Tool name and definition properties
  - Successful invocation returning trace.info
  - Edge cases with None traces and missing trace.info
  - Different trace states (OK, ERROR)
- Total 13 tests providing minimal essential coverage
- All tests pass with proper TraceInfo construction using correct field names

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Simplify builtin tool registration in judge tool registry

- Remove unnecessary _register_builtin_tools() function
- Register GetTraceInfoTool directly at module bottom
- Maintains same functionality with cleaner code
- Import still at end to avoid circular imports

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Clean up test files: remove useless docstrings and inline comments

- Remove module-level docstrings from both test files
- Remove function-level docstrings (function names are self-explanatory)
- Remove obvious inline comments like '# Create trace' and '# Invoke tool'
- Remove MockTool class docstring
- Keep code concise and focused on actual test logic
- All 13 tests continue to pass with same coverage

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove useless inline comments from registry.py

- Remove '# Get the tool from the registry' comment
- Remove '# Parse the JSON arguments' comment
- Remove '# Invoke the tool with the trace and parsed arguments' comment
- Remove '# Global registry instance' comment
- Remove '# Register built-in tools - import at the end...' comment
- Keep code clean and self-explanatory without obvious comments

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove overly defensive programming and improve docstrings

- Simplify GetTraceInfoTool.invoke() to just return trace.info directly
- Remove unnecessary null checks - assume valid trace and trace.info
- Change return type from TraceInfo | None to TraceInfo
- Update tests to remove None-related test cases
- Improve name property docstring in JudgeTool base class
- Cleaner, more direct implementation without unnecessary defensive code

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Simplify judge tool registry by removing abstract base class

- Remove JudgeToolStore abstract base class
- Merge JudgeToolStore functionality directly into JudgeToolRegistry
- Update JudgeToolRegistry to manage tools directly
- Maintain all existing functionality with simplified architecture
- Reduce unnecessary abstraction layer and improve code clarity

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Fix missing newline at end of test file

- Add newline at end of test_judge_tool_registry.py
- Fixes ruff W292 error

Signed-off-by: dbczumar <corey.zumar@databricks.com>

mlflow-app[bot] committed:
Autoformat: https://github.com/mlflow/mlflow/actions/runs/17248662714

Signed-off-by: mlflow-app[bot] <mlflow-app[bot]@users.noreply.github.com>

BenWilson2 commented on mlflow/genai/judges/tools/registry.py:
```diff
@@ -0,0 +1,124 @@
+            tool: The JudgeTool instance to register
+        """
+        self._tools[tool.name] = tool
+
```
Do we want to add a de-register method to this class?

BenWilson2 replied:
Even though this is purely backend, I'm wondering if a purger might be helpful in case a particular tool fails too many times and it can be removed as 'not applicable for use for this judge'?

BenWilson2 commented on tests/genai/judges/test_judge_tool_registry.py:
```diff
@@ -0,0 +1,192 @@
+from mlflow.utils.annotations import experimental
+
+
+@experimental(version="3.4.0")
```
Do we need to do this in tests?

BenWilson2 commented on tests/genai/judges/test_judge_tool_registry.py:
```diff
@@ -0,0 +1,192 @@
+def test_global_functions_work():
+    from mlflow.genai.judges.tools.registry import _judge_tool_registry
+
+    original_tools = _judge_tool_registry.list_tools()
```
Ensuring that it's empty or that it isn't mutated at the end of the test?

dbczumar replied:
Done! :D

BenWilson2 reviewed (approved):
Looks great! Tiny nits and some general non-blocking questions

dbczumar committed:
Address PR review comments for judge tool registry

- Add constants.py file with ToolNames class for centralized tool name references
- Update GetTraceInfoTool to use ToolNames.GET_TRACE_INFO constant
- Remove @experimental decorator from MockTool class in tests
- Add explanatory comment for original_tools tracking in test
- Simplify registry by not including deregister functionality for now

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar commented on mlflow/genai/judges/tools/get_trace_info.py:
```diff
@@ -0,0 +1,61 @@
+
+    @property
+    def name(self) -> str:
+        return "get_trace_info"
```
Absolutely - done!

dbczumar committed:
Remove useless original_tools tracking from test

- Remove unnecessary original_tools variable and related code
- Simplify test by removing pointless state tracking
- Test now directly verifies functionality without redundant checks

Signed-off-by: dbczumar <corey.zumar@databricks.com>

mlflow-app[bot] committed:
Merge remote-tracking branch 'base/judges' into tool_registry_for_judges

mlflow-app[bot] committed:
Autoformat: https://github.com/mlflow/mlflow/actions/runs/17250838710

Signed-off-by: mlflow-app[bot] <mlflow-app[bot]@users.noreply.github.com>

dbczumar commented:
Failures are unrelated. Merging to feature branch.
